stellargraph/layer/node2vec.py

```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+node2vec
+
+"""
+__all__ = ["Node2Vec"]
+
+from tensorflow.keras import Input
+from tensorflow.keras.layers import Reshape, Embedding
+import math
+from tensorflow import keras
+import warnings
+from .misc import deprecated_model_function
+from ..mapper import Node2VecLinkGenerator, Node2VecNodeGenerator
+
+
+class Node2Vec:
+    """
+    Implementation of the node2vec algorithm of A. Grover and J. Leskovec with Keras layers.
+    see: https://snap.stanford.edu/node2vec/
+
+    The model minimally requires specification of the embedding size and a generator object.
+
+    Args:
+        emb_size (int): The dimension of node embeddings.
+        generator (Sequence): A NodeSequence or LinkSequence.
+        **kwargs: Parameters node_num(int) and multiplicity(int).
+    """
+
+    def __init__(self, emb_size, generator=None, **kwargs):
+
+        # Get the node_num from the generator if it is given
+        self.generator = generator
+        if generator is not None:
+            self._get_sizes_from_generator(generator)
+        else:
+            self._get_sizes_from_keywords(kwargs)
+
+        # Model parameters
+        self.emb_size = emb_size
+
+        # Initialise the embedding layers: input-to-hidden and hidden-to-output
+        input_initializer = keras.initializers.RandomUniform(minval=-1.0, maxval=1.0)
+        self.input_embedding = Embedding(
+            self.input_node_num,
+            self.emb_size,
+            input_length=1,
+            name="input_embedding",
+            embeddings_initializer=input_initializer,
+        )
+
+        output_initializer = keras.initializers.TruncatedNormal(
```

Comment:
should this initializer be exposed as a parameter, similar to our other models?

(similarly for the `input_initializer`)

Reply:
I set the initializers by referring to the word2vec C and tensorflow implementations, which guarantees the good experimental results. Word2vec is sensitive to the initialization, other settings might fail to perform well.